### ISSUE TITLE
Error Invalid appsecret_proof when posting to facebook page

### ISSUE THREAD

**Issue by popnoart** [hybridauth/hybridauth#910]
Hi!
I'm getting this error when trying to post to a Facebook Page:
`Invalid appsecret_proof provided in the API argument`

Login its ok, post to user status also Ok, but fails with pages. Key and secret are both fine, and  "Require App Secret" is disabled.

Hybrid Auth version: hybridauth-3.0.0-rc.1
Provider: Facebook

Any ideas? Thanks!
This is the code I'm using for testing:
```
include 'Hybridauth/autoload.php';
use Hybridauth\Hybridauth; 

$config = [
    'callback' => 'MYCALLBACK',
    'keys' => [
        'id'     => 'MYFBKEY', 
        'secret' => 'MYFBSECRET' 
    ]
];

    $adapter = new \Hybridauth\Provider\Facebook($config);
    $adapter->authenticate();
    $accessToken = $adapter->getAccessToken();
    $isConnected = $adapter->isConnected();
 
    $facebook_obj = $adapter->apiRequest('/me/accounts',$accessToken['access_token']);
    $facebook_array = json_decode(json_encode($facebook_obj->data), true);

     foreach ($facebook_array as $page_key => $page_data) {
        if($page_data['id']=='MYPAGEID'){	
            $params = array(
                'access_token' => $page_data['access_token'],
                'message' => 'Prueba', 
            );
            $apiResponse = $adapter->apiRequest('/MYPAGEID/feed', 'POST', $params);
                                            
        }
    }
 

    //Disconnect the adapter 
    $adapter->disconnect();

```

**Comment by ApacheEx** [comment]
Hi,

from your example:

1.
> $facebook_obj = $adapter->apiRequest('/me/accounts',$accessToken['access_token']);

The second argument in `apiRequest` is a request method. Btw, you don't need to put access token here. It's done automatically by Hybridauth.

2.
```
$params = array(
  'access_token' => $page_data['access_token'],
  'message' => 'Prueba', 
);
$apiResponse = $adapter->apiRequest('/MYPAGEID/feed', 'POST', $params);
```
the same here, you don't need to put `access_token` here.

pls adap your code and try once more.

**Comment by popnoart** [comment]
Hi, thanks!

No luck with the new code!
Error thrown:

```
Fatal error:  Uncaught Hybridauth\Exception\HttpRequestFailedException: Signed API request has returned an error. HTTP error 400. Raw Provider API response: {"error":{"message":"Invalid appsecret_proof provided in the API argument","type":"GraphMethodException","code":100,"fbtrace_id":"GYkaC+yNe2p"}}. in /var/www/fandoms/fandoms.com/public/lab/HA/Hybridauth/Adapter/AbstractAdapter.php:344
Stack trace:
#0 /var/www/fandoms/fandoms.com/public/lab/HA/Hybridauth/Adapter/OAuth2.php(682): Hybridauth\Adapter\AbstractAdapter->validateApiResponse('Signed API requ...')
#1 /var/www/fandoms/fandoms.com/public/lab/HA/HA3.php(21): Hybridauth\Adapter\OAuth2->apiRequest('https://graph.f...')
#2 {main}
  thrown in /var/www/fandoms/fandoms.com/public/lab/HA/Hybridauth/Adapter/AbstractAdapter.php on line 344

```

Code:
```

include 'Hybridauth/autoload.php';

use Hybridauth\Hybridauth; 

$config = [
    'callback' => 'MYCALLBACK',
    'keys' => [
        'id'     => 'MYFBKEY', 
        'secret' => 'MYFBSECRET' 
    ]
];

    $adapter = new \Hybridauth\Provider\Facebook($config);
    $adapter->authenticate();
    $accessToken = $adapter->getAccessToken();    
    $isConnected = $adapter->isConnected(); 
    
    $facebook_obj = $adapter->apiRequest('/me/accounts');
    $facebook_array = json_decode(json_encode($facebook_obj->data), true);
    foreach ($facebook_array as $page_key => $page_data) {
        if($page_data['id']=='MYPAGEID'){	
            $params = array(
                'message' => 'Prueba', 
            );
            $apiResponse = $adapter->apiRequest('/MYPAGEID/feed', 'POST', $params);
        }
    }
    $adapter->disconnect();
```

**Comment by ApacheEx** [comment]
I see a bug in your code.


> $apiResponse = $adapter->apiRequest('/MYPAGEID/feed', 'POST', $params);
it should be like this:

```php
$apiResponse = $adapter->apiRequest('/MYPAGEID/feed', $params, 'POST');
```

can you check?


p.s I'm going to add needed methods to Facebook provider - so, post to page will work out of box.

**Comment by popnoart** [comment]
That would be awesome! Thanks!

**Comment by ApacheEx** [comment]
OK, it will be merged soon in master, then you can check how it works.

here is small example:

```php
$config = [
    'callback' => 'MYCALLBACK',
    'scope' => 'email, public_profile, publish_actions, manage_pages', // "manage_pages" is required.
    'keys' => [
        'id'     => 'MYFBKEY', 
        'secret' => 'MYFBSECRET' 
    ],
];

$adapter = new \Hybridauth\Provider\Facebook($config);
$adapter->authenticate();

// Will post to page wall from Page account.
$response = $adapter->setPageStatus('Hello World!', 'MYPAGEID');

// Will post to page wall from logged-in User account.
$response = $adapter->setUserStatus('Hello World!', 'MYPAGEID');

$adapter->disconnect();
```

**Comment by popnoart** [comment]
Perfect with just a small addition in scopes! For setPageStatus to work for me I had to ask for "publish_pages" too in scopes. If not throws this error:

`Requires either publish_actions permission, or manage_pages and publish_pages as an admin with sufficient administrative permission`

**Comment by ApacheEx** [comment]
Good to know, thanks. It's ok to have such errors, because if you want to use more features you need to add more scope 👌

**Comment by yaacine** [comment]
I had to deal with the same issue while working with `passport-facebook-token`, 
I finally released that the problem had nothing to have with the logic of my codebase or the app configuration.

I had this error just because I was adding intentionally an authorization Header to the request. so if you are using postman or some other http client just make sure that the request does not contain any authorization Header.
